### PR TITLE
Implement `searchModelVersions()` API in Java client

### LINKED ISSUE #6615: [FR] Implement a searchModelVersions API in the MLflow Java client
### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

Implement a searchModelVersions API in the MLflow Java client with equivalent behavior to https://github.com/mlflow/mlflow/blob/0645f9228b874e0b1d9f37cb0a0761c0736fea30/mlflow/tracking/client.py#L2700.

### Motivation

> #### What is the use case for this feature?

> #### Why is this use case valuable to support for MLflow users in general?

> #### Why is this use case valuable to support for your project(s) or organization?

> #### Why is it currently difficult to achieve this use case?

### Details

For reference, the existing implementation of searchExperiments may be helpful because the searchModelVersions implementation is likely to be very similar: https://github.com/mlflow/mlflow/blob/0645f9228b874e0b1d9f37cb0a0761c0736fea30/mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java#L275-L280

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [X] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [X] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by gabrielfu**
## Related Issues/PRs

resolves #6615 

## What changes are proposed in this pull request?
Added client-facing methods to allow search for model versions with filters. 

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Implement `searchModelVersions()` API in Java client

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [X] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java [hunk 1]
```diff
@@ -929,6 +929,33 @@ public File downloadLatestModelVersion(String modelName, String stage) {
...
+   *
+   * @param searchFilter SQL compatible search query string.
+   *                     Examples:
+   *                         - "name = 'model_name'"
+   *                         - "run_id = '...'"
+   *                     If null, the result will be equivalent to having an empty search filter.
+   *
+   * @return A page of model versions that satisfy the search filter.
+   */
+  public ModelVersionsPage searchModelVersions(String searchFilter) {
```
Comment (BenWilson2) [inline mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java:952]: could we update this with the recently added `order_by`, `max_results`, and `page_token` arguments that the Python API now supports?

Comment (BenWilson2) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java:952]: Updates can be seen in this PR: https://github.com/mlflow/mlflow/pull/7623

File: mlflow/java/client/src/test/java/org/mlflow/tracking/ModelRegistryMlflowClientTest.java [hunk 2]
```diff
@@ -160,4 +160,45 @@ private void validateDetailedModelVersion(ModelVersion details, String modelName
...
+        Assert.assertEquals(mvs2.size(), 1);
+        Assert.assertEquals(mvs2.get(0).getName(), modelName2);
+        Assert.assertEquals(mvs2.get(0).getVersion(), "1");
+
+        String filter3 = String.format("run_id = '%s'", newVersionRunId);
+        List<ModelVersion> mvs3 = client.searchModelVersions(filter3).getItems();
+        Assert.assertEquals(mvs3.size(), 1);
+        Assert.assertEquals(mvs3.get(0).getName(), modelName);
+        Assert.assertEquals(mvs3.get(0).getVersion(), "2");
+    }
```
Comment (BenWilson2) [inline mlflow/java/client/src/test/java/org/mlflow/tracking/ModelRegistryMlflowClientTest.java:203]: Once adding in `orderby` and the continuation token support, can we add tests to validate ordering and paged requests?

File: mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java (hunk 1 above)
Comment (gabrielfu) [reply mlflow/java/client/src/main/java/org/mlflow/tracking/MlflowClient.java:952]: Oh I missed it! Of course, let me update this.

File: mlflow/java/client/src/test/java/org/mlflow/tracking/ModelRegistryMlflowClientTest.java (hunk 2 above)
Comment (gabrielfu) [reply mlflow/java/client/src/test/java/org/mlflow/tracking/ModelRegistryMlflowClientTest.java:203]: Sure!

**Review by BenWilson2 — APPROVED** [review]
This is great! Thank you for the implementation and the super fast responses!

**Comment by BenWilson2** [comment]
Hey @gabrielfu could you check to see if you have any issues in your fork related to the circleci config file? Since the Java API has been changed, we just need to make sure that the Java Docs will render properly in order to merge.

**Comment by gabrielfu** [comment]
> Hey @gabrielfu could you check to see if you have any issues in your fork related to the circleci config file? Since the Java API has been changed, we just need to make sure that the Java Docs will render properly in order to merge.

Hi, not sure what's happening with circleci, but I can see the docs locally:
image
image
image

**Comment by BenWilson2** [comment]
Thank you for the screenshot and confirming the correct rendering!

**Pull request merged**